From c34229d87304c7f7f5713c0e34c001ab534ae384 Mon Sep 17 00:00:00 2001
From: Thao Luong <thao.luong.uw@rvc.renesas.com>
Date: Wed, 27 May 2015 11:33:29 +0700
Subject: [PATCH 14/25] omxvideodec: Fix "not-negotiated" error due to 
 pixel-aspect-ratio not 1/1

This error happens because v4l2sink plugin does not support other
pixel-aspect-ratio than 1/1

refs #69206
Signed-off-by: Thao Luong <thao.luong.uw@rvc.renesas.com>
---
 config/rcar/gstomx.conf | 2 +-
 omx/gstomx.c            | 2 ++
 omx/gstomx.h            | 6 ++++++
 omx/gstomxvideodec.c    | 6 ++++++
 4 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/config/rcar/gstomx.conf b/config/rcar/gstomx.conf
index 5f859f5..d93362e 100644
--- a/config/rcar/gstomx.conf
+++ b/config/rcar/gstomx.conf
@@ -5,7 +5,7 @@ component-name=OMX.RENESAS.VIDEO.DECODER.H264
 rank=512
 in-port-index=0
 out-port-index=1
-hacks=no-disable-outport
+hacks=no-disable-outport;default-pix-aspect-ratio
 
 [omxaacdec]
 type-name=GstOMXAACDec
diff --git a/omx/gstomx.c b/omx/gstomx.c
index 59b1dc0..5a4c242 100644
--- a/omx/gstomx.c
+++ b/omx/gstomx.c
@@ -2464,6 +2464,8 @@ gst_omx_parse_hacks (gchar ** hacks)
       hacks_flags |= GST_OMX_HACK_NO_DISABLE_OUTPORT;
     else if (g_str_equal (*hacks, "not-handle-codec-data"))
       hacks_flags |= GST_OMX_HACK_NOT_HANDLE_CODEC_DATA;
+    else if (g_str_equal (*hacks, "default-pix-aspect-ratio"))
+      hacks_flags |= GST_OMX_HACK_DEFAULT_PIXEL_ASPECT_RATIO;
     else
       GST_WARNING ("Unknown hack: %s", *hacks);
     hacks++;
diff --git a/omx/gstomx.h b/omx/gstomx.h
index 7051043..636a800 100644
--- a/omx/gstomx.h
+++ b/omx/gstomx.h
@@ -123,6 +123,12 @@ G_BEGIN_DECLS
  */
 #define GST_OMX_HACK_NOT_HANDLE_CODEC_DATA                            G_GUINT64_CONSTANT (0x0000000000000200)
 
+/* The videosink plugin only supports default pixel-aspect-ratio
+ * (1/1), cannot support other pixel-aspect-ratio
+ * Happens with Renesas Gen2 platfrom
+ */
+#define GST_OMX_HACK_DEFAULT_PIXEL_ASPECT_RATIO                       G_GUINT64_CONSTANT (0x0000000000000400)
+
 typedef struct _GstOMXCore GstOMXCore;
 typedef struct _GstOMXPort GstOMXPort;
 typedef enum _GstOMXPortDirection GstOMXPortDirection;
diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index 15eba8f..0707468 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -1397,6 +1397,7 @@ gst_omx_video_dec_loop (GstOMXVideoDec * self)
   GstOMXAcquireBufferReturn acq_return;
   GstClockTimeDiff deadline;
   OMX_ERRORTYPE err;
+  GstOMXVideoDecClass *klass = GST_OMX_VIDEO_DEC_GET_CLASS (self);
 
 #if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
   port = self->eglimage ? self->egl_out_port : self->dec_out_port;
@@ -1485,6 +1486,11 @@ gst_omx_video_dec_loop (GstOMXVideoDec * self)
       gst_omx_port_update_port_definition (self->dec_out_port, NULL);
 
       /* Take framerate and pixel-aspect-ratio from sinkpad caps */
+      if (klass->cdata.hacks & GST_OMX_HACK_DEFAULT_PIXEL_ASPECT_RATIO) {
+        /* Workaround in case videosink plugin only supports
+        * default pixel-aspect-ratio 1/1   */
+        state->info.par_d = state->info.par_n;
+      }
 
       if (!gst_video_decoder_negotiate (GST_VIDEO_DECODER (self))) {
         if (buf)
-- 
2.4.0

